app/services/puppet_fact_parser.rb

```diff
@@ -2,8 +2,14 @@ class PuppetFactParser < FactParser
   attr_reader :facts
 
   def operatingsystem
-    orel = os_release
+    os=parse_hostname
+    os.save!
+    os.architectures << architecture unless os.architectures.include?(architecture) || architecture.blank?
```

Comment:
can you swap lines 6/7?  It might be a bit faster if Rails can write the association during save!.